core/translate/optimizer.rs

```diff
@@ -0,0 +1,732 @@
+    node: &mut Operator,
+    predicate: ast::Expr,
+    referenced_tables: &Vec<(Rc<BTreeTable>, String)>,
+) -> Result<Option<ast::Expr>> {
+    match node {
+        Operator::Scan {
+            predicates,
+            table_identifier,
+            ..
+        } => {
+            let table_index = referenced_tables
+                .iter()
+                .position(|(_, t_id)| t_id == table_identifier)
+                .unwrap();
+
+            let predicate_bitmask =
+                get_table_ref_bitmask_for_ast_expr(referenced_tables, &predicate)?;
+
+            // the expression is allowed to refer to tables on its left, i.e. the righter bits in the mask
+            // e.g. if this table is 0010, and the table on its right in the join is 0100:
+            // if predicate_bitmask is 0011, the predicate can be pushed (refers to this table and the table on its left)
+            // if predicate_bitmask is 0001, the predicate can be pushed (refers to the table on its left)
+            // if predicate_bitmask is 0101, the predicate can't be pushed (refers to this table and a table on its right)
+            let next_table_on_the_right_in_join_bitmask = 1 << (table_index + 1);
+            if predicate_bitmask >= next_table_on_the_right_in_join_bitmask {
+                return Ok(Some(predicate));
+            }
+
+            if predicates.is_none() {
+                predicates.replace(vec![predicate]);
+            } else {
+                predicates.as_mut().unwrap().push(predicate);
+            }
+
+            return Ok(None);
+        }
+        Operator::Filter {
+            source,
+            predicates: ps,
+            ..
+        } => {
+            let push_result = push_predicate(source, predicate, referenced_tables)?;
+            if push_result.is_none() {
+                return Ok(None);
+            }
+
+            ps.push(push_result.unwrap());
+
+            return Ok(None);
+        }
+        Operator::Join {
+            left,
+            right,
+            predicates: join_on_preds,
+            outer,
+            ..
+        } => {
+            let push_result_left = push_predicate(left, predicate, referenced_tables)?;
+            if push_result_left.is_none() {
+                return Ok(None);
+            }
+            let push_result_right =
+                push_predicate(right, push_result_left.unwrap(), referenced_tables)?;
+            if push_result_right.is_none() {
+                return Ok(None);
+            }
+
+            if *outer {
+                return Ok(Some(push_result_right.unwrap()));
+            }
+
+            let pred = push_result_right.unwrap();
+
+            let table_refs_bitmask = get_table_ref_bitmask_for_ast_expr(referenced_tables, &pred)?;
+
+            let left_bitmask = get_table_ref_bitmask_for_query_plan_node(referenced_tables, left)?;
+            let right_bitmask =
+                get_table_ref_bitmask_for_query_plan_node(referenced_tables, right)?;
```

Comment:
What is the purpose of bitmasks here? A comment might help a bit.

Reply:
mainly my paranoia about making this slow by using e.g. vecs of string references or similar. i know there's a lot of cloning elsewhere though 😂 

added comments to the relevant funcs